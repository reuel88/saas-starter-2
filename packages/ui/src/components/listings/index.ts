export { Listing } from "./listing";
export { ListingLoader } from "./listing-loader";
export { Listings } from "./listings";
