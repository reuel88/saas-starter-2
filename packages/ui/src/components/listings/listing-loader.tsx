import { FC } from "react";
import { Card, CardContent, CardFooter, CardHeader } from "../card";
import { Skeleton } from "../ui/skeleton";

type ListingLoaderProps = {};

export const ListingLoader: FC<ListingLoaderProps> = () => {
  return (
    <Card>
      <CardHeader>
        {/* Title */}
        <Skeleton className="h-6" />

        {/* Price & Location */}
        <div className="flex h-7 items-center">
          <Skeleton className="h-5 w-full" />
        </div>
      </CardHeader>

      <CardContent className="space-y-6">
        {/* Image */}
        <Skeleton className="aspect-video" />

        {/* Description */}
        <div className="space-y-1.5 pt-1.5">
          <Skeleton className="h-3.5 w-full" />
          <Skeleton className="h-3.5 w-full" />
          <Skeleton className="h-3.5 w-3/4" />
        </div>
      </CardContent>

      {/* CTA */}
      <CardFooter className="flex gap-4">
        <Skeleton className="h-10 w-20" />
        <Skeleton className="h-10 w-20" />
      </CardFooter>
    </Card>
  );
};
