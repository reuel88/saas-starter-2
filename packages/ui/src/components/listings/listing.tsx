import { FC, ReactNode } from "react";
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from "../card";
import { Button } from "../button";

export type ListingProps = {
  title: string;
  price: number;
  location?: string;
  imageSrc?: string;
  description: ReactNode;
};
export const Listing: FC<ListingProps> = ({
  title,
  price,
  location,
  imageSrc,
  description,
  ...rest
}) => {
  return (
    <Card as="article" className="flex flex-col" {...rest}>
      <CardHeader as="header">
        <CardTitle className="line-clamp-1">{title}</CardTitle>
        <div className="flex items-center justify-between gap-4">
          <p className="text-lg font-extrabold text-hue-color">
            {new Intl.NumberFormat("en-AU", {
              style: "currency",
              currency: "AUD",
              minimumFractionDigits: 0,
            }).format(price)}
          </p>
          {location && (
            <p className="text-sm text-muted-foreground">{location}</p>
          )}
        </div>
      </CardHeader>

      <CardContent className="space-y-6">
        {imageSrc && (
          <img
            src={imageSrc}
            alt={title}
            className="aspect-video object-cover"
          />
        )}

        <CardDescription className="line-clamp-3">
          {description}
        </CardDescription>
      </CardContent>

      <CardFooter as="footer" className="mt-auto flex gap-4">
        <Button className="min-w-20" type="button">
          View
        </Button>
        <Button variant="outline" className="min-w-20" type="button">
          Reply
        </Button>
      </CardFooter>
    </Card>
  );
};
