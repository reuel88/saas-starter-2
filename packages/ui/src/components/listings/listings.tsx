import { FC } from "react";
import { Listing, ListingProps } from "./listing";

type ListingsProps = {
  listings: ListingProps[];
};
export const Listings: FC<ListingsProps> = ({ listings }) => {
  return (
    <section>
      <header className="border-b py-4">
        <h2 className="text-xl">Search Results</h2>
        <h3 className="text-muted-foreground">
          <strong className="text-hue-color">10 results</strong> for{" "}
          <strong className="text-hue-color">Cars</strong> in{" "}
          <strong className="text-hue-color">Locations</strong>
        </h3>
      </header>
      <div className="grid grid-cols-2 gap-4 py-8 lg:grid-cols-3">
        {listings.map((listing, index) => (
          <Listing key={index} {...listing} data-testid="listing" />
        ))}
      </div>
    </section>
  );
};
