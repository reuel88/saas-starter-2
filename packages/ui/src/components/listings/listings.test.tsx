import { render, screen } from "@testing-library/react";
import { Listings } from "./listings";

describe("Listings", () => {
  it("should render search results header", () => {
    // Arrange

    // Act
    const { baseElement } = render(<Listings listings={[]} />);
    const header = screen.getByRole("heading", {
      level: 2,
      name: /search results/i,
    });

    // Assert
    expect(header).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render correct number of listings", () => {
    // Arrange
    const listings = [
      { id: 1, title: "Car 1", price: 10000, description: "Description 1" },
      { id: 2, title: "Car 2", price: 1000, description: "Description 2" },
      { id: 3, title: "Car 3", price: 100, description: "Description 3" },
    ];

    // Act
    render(<Listings listings={listings} />);
    const listingElements = screen.getAllByTestId("listing");

    // Assert
    expect(listingElements.length).toBe(listings.length);
  });
});
