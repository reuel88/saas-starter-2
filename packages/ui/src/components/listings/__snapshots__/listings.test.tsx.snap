// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Listings > should render search results header 1`] = `
<body>
  <div>
    <section>
      <header
        class="border-b py-4"
      >
        <h2
          class="text-xl"
        >
          Search Results
        </h2>
        <h3
          class="text-muted-foreground"
        >
          <strong
            class="text-hue-color"
          >
            10 results
          </strong>
           for
           
          <strong
            class="text-hue-color"
          >
            Cars
          </strong>
           in
           
          <strong
            class="text-hue-color"
          >
            Locations
          </strong>
        </h3>
      </header>
      <div
        class="grid grid-cols-2 gap-4 py-8 lg:grid-cols-3"
      />
    </section>
  </div>
</body>
`;
